File: 剑指Offer48-zui-chang-bu-han-zhong-fu-zi-fu-de-zi-zi-fu-chuan-lcof/48.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        map<char, int> m;
        int l = -1, r = 0, res = 0;
        while (r < s.length())
        {
            if (m.count(s[r]))
            {
                l = max(l, m[s[r]]);
            }
            m[s[r]] = r;
            res = max(res, r - l);
            ++r;
        }
        return res;
    }
};

int main(int argc, char **args)
{
    Solution sol = Solution();
    cout << sol.lengthOfLongestSubstring("abcabcbb") << endl;
    cout << sol.lengthOfLongestSubstring("bbbbb") << endl;
    cout << sol.lengthOfLongestSubstring("pwwkew") << endl;
    cout << sol.lengthOfLongestSubstring(" ") << endl;
    return 0;
}

File: 剑指Offer48-zui-chang-bu-han-zhong-fu-zi-fu-de-zi-zi-fu-chuan-lcof/48.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
from typing import Dict


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dic: Dict[str, int] = {}
        l, res = -1, 0
        for r in range(len(s)):
            if s[r] in dic:
                l = max(l, dic[s[r]])

            dic[s[r]] = r
            res = max(res, r - l)
        return res


if __name__ == '__main__':
    sol = Solution()
    print(sol.lengthOfLongestSubstring("abcabcbb"))
    print(sol.lengthOfLongestSubstring("bbbbb"))
    print(sol.lengthOfLongestSubstring("pwwkew"))
    print(sol.lengthOfLongestSubstring(" "))
